backend/cn/main.ml

```diff
@@ -246,6 +234,54 @@ let main
       Cerb_debug.maybe_close_csv_timing_file_no_err ();
       Printexc.raise_with_backtrace exc (Printexc.get_raw_backtrace ())
 
+let generate_tests
+      (* Common *)
+      filename
+      macros
+      incl_dirs
+      incl_files
+      astprints
+      use_peval
+      no_inherit_loc
+      magic_comment_char_dollar
+      (* Test Generation *)
+      output_dir
+      max_unfolds
+      testing_framework
+  =
+  check_input_file filename;
+  let (prog4, (markers_env, ail_prog), _) =
+    handle_frontend_error
+      (frontend ~macros ~incl_dirs ~incl_files astprints ~do_peval:use_peval ~filename ~magic_comment_char_dollar)
+  in
+  let (_, sigma) = ail_prog in
+  try
+    let handle_error (e : TypeErrors.type_error) =
+      let report = TypeErrors.pp_message e.msg in
+      Pp.error e.loc report.short (Option.to_list report.descr);
+      print_string ("\nRun `cn verify " ^ filename ^ "` for more details\n");
+      match e.msg with
+      | TypeErrors.Unsupported _ -> exit 2
+      | _ -> exit 1
+    in
+    let result =
+      let open Resultat in
+      let@ prog5 = Core_to_mucore.normalise_file ~inherit_loc:(not(no_inherit_loc)) (markers_env, snd ail_prog) prog4 in
+      print_log_file ("mucore", MUCORE prog5);
+      let paused = Typing.run_to_pause Context.empty (Check.check_decls_lemmata_fun_specs prog5) in
+      Result.iter_error handle_error (Typing.pause_to_result paused);
+      Cerb_colour.without_colour (fun () ->
+        TestGeneration.main ~output_dir ~filename ~max_unfolds sigma prog5 testing_framework)
+      ();
+      return ()
+    in
+    Pp.maybe_close_times_channel ();
+    Result.fold ~ok:(fun () -> exit 0) ~error:handle_error result
+  with
+  | exc ->
+    Pp.maybe_close_times_channel ();
+    Cerb_debug.maybe_close_csv_timing_file_no_err ();
+    Printexc.raise_with_backtrace exc (Printexc.get_raw_backtrace ())
```

Comment:
This is a lot of code duplication. Is it possible to factor out the bits common to the existing `main` and generate tests?

Reply:
Uhh, so I did this, but it's not the most beautiful thing, so I can revert it.

A side effect is that it was trivial to add a `cn check` which just does the well-formedness check (I was thinking along the lines of `cargo check`, though I can remove it)

Reply:
I think `cn lint-spec` might be more accurate since `check` and `verify` sound a little too similar.

Reply:
But `lint-spec` is not what it does. Perhaps `cn wellformed`, even if that's not a verb.

Reply:
What about `cn sanity-check`?